x/programs/rust/examples/multisig/src/lib.rs

```diff
@@ -0,0 +1,495 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use wasmlanche_sdk::{
+    public, state_keys, Address, Context, DeferDeserialize, ExternalCallError, Program,
+};
+
+pub const MIN_VOTES: u64 = 2;
+
+#[derive(Debug, BorshSerialize, BorshDeserialize)]
+pub struct Proposal {
+    to: Program,
+    method: String,
+    args: Vec<u8>,
+    value: u64,
+    executed: bool,
+    voters_len: u64,
+}
+
+#[derive(Debug, BorshSerialize, BorshDeserialize)]
+pub struct Voter {
+    address: Address,
+    voted: bool,
+}
+
+#[derive(Debug, BorshSerialize, BorshDeserialize)]
+pub enum ProposalError {
+    AlreadyExecuted,
+    AlreadyVoted,
+    InexistentProposal,
+    QuorumNotReached,
+    NotVoter,
+    ExecutionFailed(ExternalCallError),
+}
+
+#[state_keys]
+pub enum StateKeys {
+    LastProposalId,
+    Proposals(u64),
+    Voter(u64, usize),
+    Yeas(u64),
+    Nays(u64),
+}
+
+#[public]
+pub fn propose(
+    context: Context<StateKeys>,
+    voters_addr: Vec<Address>,
+    to: Program,
+    method: String,
+    args: Vec<u8>,
+    value: u64,
+) -> u64 {
+    let voters_len = voters_addr.len().try_into().expect("too much voters");
+    assert!(voters_len >= MIN_VOTES);
+    assert!(_is_not_duplicate(&voters_addr));
+    let mut voters: Vec<_> = voters_addr
+        .iter()
+        .map(|address| Voter {
+            address: *address,
+            voted: false,
+        })
+        .collect();
+    let i = _ensure_voter(&context.actor(), &voters).unwrap();
+    voters[i as usize].voted = true;
+
+    let last_id = _last_proposal_id(&context);
+
+    context
+        .store_by_key(
+            StateKeys::Proposals(last_id),
+            &Proposal {
+                to,
+                method,
+                args,
+                value,
+                executed: false,
+                voters_len,
+            },
+        )
+        .unwrap();
+
+    for (i, voter) in voters.into_iter().enumerate() {
+        context
+            .store_by_key(StateKeys::Voter(last_id, i), &voter)
+            .unwrap();
+    }
+
+    context
+        .store([
+            (StateKeys::Nays(last_id), &0u64),
+            (StateKeys::Yeas(last_id), &1),
+        ])
+        .unwrap();
+
+    context
+        .store_by_key(StateKeys::LastProposalId, &(last_id + 1))
+        .unwrap();
+
+    last_id
+}
+
+#[public]
+pub fn vote(context: Context<StateKeys>, id: u64, yea: bool) -> Result<(), ProposalError> {
+    let proposal = _proposal_at(&context, id).ok_or(ProposalError::InexistentProposal)?;
+
+    if proposal.executed {
+        return Err(ProposalError::AlreadyExecuted);
+    }
+
+    let actor = context.actor();
+    let voters: Vec<_> = (0..proposal.voters_len)
+        .map(|i| {
+            context
+                .get(StateKeys::Voter(id, i as usize))
+                .unwrap()
+                .unwrap()
+        })
+        .collect();
```

Comment:
This is a very inefficient way pull out all the voters, why do you need to do this? 

If you have a key `Vote(Address)`, you check if the data is there? If it is, the person can't vote again, you return an error. If it's not there, increase the yea or nay count, store the vote, done.

Reply:
Thanks, The voter now passes a `voter_id` where they expects to be in the set and the program will check if it is here and then if it has already voted.